extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -0,0 +1,654 @@
+  {
+    'name': 'OnePlus 5T',
+    'width': 455,
+    'height': 820,
+    'deviceHeight': 910,
+  },
+  {
+    'name': 'OnePlus 7 Pro',
+    'width': 412,
+    'height': 743,
+    'deviceHeight': 892,
+  },
+  {
+    'name': 'Desktop 1080p',
+    'width': 1920,
+    'height': 1080,
+    'deviceHeight': 1080,
+  },
+];
+
+/**
+ * Method that returns lowercase and remove non-alphanumeric, used on matching hashString.
+ * @param {string} name
+ * @return {string}
+ */
+function simplifyDeviceName(name) {
+  return name.toLowerCase().replace(/[^a-z0-9]/gi, '');
+}
+
+/**
+ * Get devices from the queryHash into list of devices
+ * @param {string} queryHash
+ * @return {any[]}
+ */
+function parseDevices(queryHash) {
+  const screenSizes = [];
+
+  queryHash.split(';').forEach((device) => {
+    const deviceSpecs = device.split(':');
+    let currSpecs = null;
+    if (deviceSpecs.length == 1) {
+      currSpecs = ALL_SCREEN_SIZES.find((el) => {
+        // Find first device that has prefix of the device name passed in.
+        const currDeviceName = simplifyDeviceName(el.name);
+        const specDeviceName = simplifyDeviceName(deviceSpecs[0]);
+        return (
+          currDeviceName.substring(0, specDeviceName.length) == specDeviceName
+        );
+      });
+    }
+    if (currSpecs) {
+      screenSizes.push(currSpecs);
+    }
+  });
+  return screenSizes;
+}
+
+/**
+ * @typedef {{
+ *  element: !Element,
+ *  player: !Element
+ *  chip: !Element,
+ *  width: number,
+ *  height: number,
+ *  deviceHeight: ?number,
+ *  custom: ?boolean,
+ * }}
+ */
+export let DeviceInfo;
+
+export class AmpStoryDevToolsTabPreview extends AMP.BaseElement {
+  /** @param {!Element} element */
+  constructor(element) {
+    super(element);
+
+    /** @private  {?string} */
+    this.storyUrl_ = null;
+
+    /** @private {!Array<DeviceInfo>} */
+    this.devices_ = [];
+  }
+
+  /** @override */
+  buildCallback() {
+    const styleTag = this.element.ownerDocument.createElement('style');
+    styleTag.textContent = CSS;
+    this.element.appendChild(styleTag);
+
+    this.storyUrl_ = this.element.getAttribute('story-url');
+    this.element.classList.add('i-amphtml-story-dev-tools-tab');
+    const chipListContainer = this.element.ownerDocument.createElement('div');
+    chipListContainer.classList.add(
+      'i-amphtml-story-dev-tools-device-chips-container'
+    );
+    this.element.appendChild(chipListContainer);
+
+    const chipList = this.element.ownerDocument.createElement('span');
+    chipList.classList.add('i-amphtml-story-dev-tools-device-chips');
+    chipListContainer.appendChild(chipList);
+    chipList.addEventListener('click', (event) => {
+      const chip = closest(event.target, (el) => el.device);
+      if (chip) {
+        this.removeDevice_(chip.device);
+        this.updateDevicesInHash_();
+      }
+    });
+
+    chipListContainer.appendChild(this.buildAddDeviceButton_());
+    chipListContainer.appendChild(this.buildHelpButton_());
+
```

Comment:
Nit: event delegation

FWIW: Most high performance apps have only one click event listener at the `window` level that capture all clicks. This architecture has a ton of benefits that we unfortunately can't use here but it's always good to keep the event delegation as a good practice, it makes your event listeners logic much easier to understand and maintain. If each button has its own listeners and you start playing around with `capture` and `stopPropagation` as most apps do, it becomes extremely messy and impossible to predict.